crates/kitsune_p2p/proxy/src/tls_cli.rs

```diff
@@ -0,0 +1,142 @@
+        send,
+        recv,
+        write,
+        read,
+    ));
+    setup_recv
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn tls_client(
+    short: String,
+    setup_send: tokio::sync::oneshot::Sender<TransportResult<()>>,
+    expected_proxy_url: ProxyUrl,
+    tls_client_config: Arc<rustls::ClientConfig>,
+    mut send: TransportChannelWrite,
+    recv: TransportChannelRead,
+    mut write: futures::channel::mpsc::Sender<ProxyWire>,
+    read: futures::channel::mpsc::Receiver<ProxyWire>,
+) {
+    let mut setup_send = Some(setup_send);
+    let res: TransportResult<()> = async {
+        let nr = webpki::DNSNameRef::try_from_ascii_str("stub.stub").unwrap();
+        let mut cli = rustls::ClientSession::new(&tls_client_config, nr);
+        let mut buf = [0_u8; 4096];
+        let mut in_pre = std::io::Cursor::new(Vec::new());
+
+        let mut merge = kitsune_p2p_types::auto_stream_select(recv, read);
+        use kitsune_p2p_types::AutoStreamSelect::*;
+
+        let mut wants_write_close = false;
+        let mut did_post_handshake_work = false;
+        loop {
+            if !did_post_handshake_work && !cli.is_handshaking() {
+                did_post_handshake_work = true;
+
+                let cert_digest = blake2b_32(
+                    cli.get_peer_certificates()
+                        .unwrap()
+                        .get(0)
+                        .unwrap()
+                        .as_ref(),
+                );
+
+                let remote_proxy_url =
+                    ProxyUrl::new(expected_proxy_url.as_base().as_str(), cert_digest.into())?;
+                if let Some(setup_send) = setup_send.take() {
+                    if expected_proxy_url == remote_proxy_url {
+                        tracing::info!("{}: CLI: CONNECTED TLS: {}", short, remote_proxy_url);
+                        let _ = setup_send.send(Ok(()));
+                    } else {
+                        let msg = format!(
+                            "expected remote {} != received {}",
+                            expected_proxy_url, remote_proxy_url,
+                        );
+                        let _ = setup_send.send(Err(msg.clone().into()));
+                        return Err(msg.into());
+                    }
+                }
+            }
+
+            if cli.wants_write() {
+                tracing::trace!("{}: CLI tls wants write", short);
+                let mut data = Vec::new();
+                cli.write_tls(&mut data).map_err(TransportError::other)?;
+                write
+                    .send(ProxyWire::chan_send(data.into()))
+                    .await
+                    .map_err(TransportError::other)?;
+            } else if wants_write_close && !cli.is_handshaking() {
+                write.close().await.map_err(TransportError::other)?;
+            }
+
+            if !cli.wants_read() {
+                tokio::time::delay_for(std::time::Duration::from_millis(10)).await;
```

Comment:
How come this needs to loop with a delay? Can it not await on the read and write streams?

Reply:
This is a part of the mechanics of the not async-aware rustls API - we don't actually have anything to wait on here...

It's entirely possible this is not implemented correctly and/or there is a better way to do it, but honestly, I finally got something to work and didn't want to muck with it any more right now *blush* : )